### PR TITLE
[CAS-1266] Refactored watch channel to use LLC and OfflinePlugin

### PR DESCRIPTION
**Pull request by ogkuzmin**
### 🎯 Goal

We have `ChannelController::runChannelQuery` which has several side effects and perform queries using `ChatClient`. By merging ChatDomain into LLC we want to eliminate additional entities (`ChannelController`, `ChatDomain`) and be able to get the same behavior.

### 🛠 Implementation details

1. Introduce `precondition` callback in `Plugin`. Which returns `Result.success(Unit)` if precondition is satisfied and `Result.error` otherwise. Now we can rule the flow of LLC requests. It's needed because in offline support we don't fire a request, if the previous one is still in progress.
2. Enrich `QueryChannelRequest` to get data of limits, offsets, and pagination
3. Added some KDOCs for new changes.
4. Moved StateFlow and variables from `ChannelController` to `ChannelMutableState`
5. Moved side effects of query channels from `ChannelController` to `ChannelLogic`
6. Kept backward compatibility to wrap state and logic in `ChannelController`
7. Introduced `QueryChannelReference` as wrapper for `ChatClient::queryChannel`
8. Added `ChatClientReferenceAdapter::queryChannel` and `ChatClientReferenceAdapter::watchChannel`

### 🧪 Testing

1. Substitude channelController usage by new state in `MessageListViewMolde`
2. Run tests (we had some for `ChannelController`)

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [x] Affected documentation updated (docusaurus, tutorial, CMS (task created))
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt [hunk 1]
```diff
@@ -53,42 +50,7 @@ internal class ChannelControllerImpl(private val channelControllerStateFlow: Cha
     override var recoveryNeeded = channelControllerStateFlow.recoveryNeeded
     override val cid = channelControllerStateFlow.cid
 
-    fun getThread(threadId: String): ThreadControllerImpl =
```
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt:56]: Unused methods, hidden from public API

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt [hunk 2]
```diff
@@ -832,78 +674,35 @@ public class ChannelController internal constructor(
...
+        if (mutableState.hideMessagesBefore != null) {
+            if (message != null && message.wasCreatedBeforeOrAt(mutableState.hideMessagesBefore)) {
                 message = null
             }
         }
 
         return message
     }
 
-    private fun parseMessages(messages: List<Message>): Map<String, Message> {
```
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt:868]: Moved these methods to `ChannelLogic`

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt [hunk 3]
```diff
@@ -1486,9 +1166,5 @@ public class ChannelController internal constructor(
         private const val KEY_MESSAGE_ACTION = "image_action"
         private const val MESSAGE_ACTION_SHUFFLE = "shuffle"
         private const val MESSAGE_ACTION_SEND = "send"
-        private const val TYPE_IMAGE = "image"
```
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt:1489]: Removed as unused

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ThreadControllerImpl.kt [hunk 4]
```diff
@@ -18,12 +16,6 @@ import io.getstream.chat.android.offline.thread.ThreadController as ThreadContro
 internal class ThreadControllerImpl(private val threadControllerStateFlow: ThreadControllerStateFlow) :
     ThreadController {
 
-    internal constructor(
```
Comment (ogkuzmin) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ThreadControllerImpl.kt:21]: Unused

File: stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/Call.kt [hunk 5]
```diff
@@ -90,6 +90,9 @@ public fun <T : Any> Call<T>.doOnStart(scope: CoroutineScope, function: suspend
 public fun <T : Any> Call<T>.doOnResult(scope: CoroutineScope, function: suspend (Result<T>) -> Unit): Call<T> =
     DoOnResultCall(this, scope, function)
 
+public fun <T: Any> Call<T>.withPrecondition(scope: CoroutineScope, precondition: suspend () -> Result<Unit>): Call<T> =
```
Comment (adasiewiczr) [inline stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/Call.kt:93]: Should we mark it at Experimental and InternalStreamApi for now?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/ChannelState.kt [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.TypingEvent
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.core.ExperimentalStreamChatApi
+import io.getstream.chat.android.offline.channel.ChannelData
+import kotlinx.coroutines.flow.StateFlow
+
+@ExperimentalStreamChatApi
+/** State container with reactive data of a channel.*/
+public interface ChannelState {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/ChannelState.kt:14]: Does it make sense to mark it with InternalApi for now?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/MessagesState.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package io.getstream.chat.android.offline.experimental.channel.state
+
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.offline.ChatDomainImpl
+
+/**
+ * Represents of possible state of messages for [ChannelState].
+ */
+public sealed class MessagesState {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/MessagesState.kt:9]: Same here. Maybe we could add @Experimental and @InternalApi for now

**Review by adasiewiczr — APPROVED** [review]

File: stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/Call.kt (hunk 5 above)
Comment (ogkuzmin) [reply stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/Call.kt:93]: I marked it with InternalStreamChatApi

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/MessagesState.kt (hunk 7 above)
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/MessagesState.kt:13]: Done

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/ChannelState.kt (hunk 6 above)
Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/state/ChannelState.kt:16]: Done

**Pull request merged**